### ISSUE TITLE
ray.init fails when temp_dir too long (but still reasonable)

### ISSUE THREAD

**Issue by zajaczajac** [ray-project/ray#7724]
### What is the problem?

When I set `temp_dir` argument of `ray.init` to a path that is still reasonably short (50 characters), exception about `File name too long` is thrown. See the code and traces below.

### Reproduction (REQUIRED)
Script:
```
import ray

ray.init(temp_dir='/tmp/aaaaaaaaaabbbbbbbbbbaaaaaaaaaabbbbbbbbbbaaaaa')
```

Trace:
```
2020-03-24 10:24:03,788	INFO resource_spec.py:212 -- Starting Ray with 4.59 GiB memory available for workers and up to 2.3 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-03-24 10:24:04,902	INFO services.py:523 -- Failed to connect to the redis server, retrying.
2020-03-24 10:24:05,234	INFO services.py:1123 -- View the Ray dashboard at localhost:8265
libc++abi.dylib: terminating with uncaught exception of type boost::wrapexcept<boost::system::system_error>: File name too long
*** Aborted at 1585041845 (unix time) try "date -d @1585041845" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x7fff6d9bf7fa) received by PID 2316 (TID 0x110b5adc0) stack trace: ***
    @     0x7fff6da7142d _sigtramp
    @              0x400 (unknown)
    @     0x7fff6d946a1c abort
    @     0x7fff6a9e0be8 abort_message
    @     0x7fff6a9e0d84 demangling_terminate_handler()
    @     0x7fff6c50c792 _objc_terminate()
    @     0x7fff6a9eddc7 std::__terminate()
    @     0x7fff6a9edb6c __cxxabiv1::failed_throw()
    @     0x7fff6a9df45d __cxa_throw
    @        0x10c889718 boost::throw_exception<>()
    @        0x10c8896ac boost::asio::detail::do_throw_error()
    @        0x10c88f5b9 boost::asio::local::detail::endpoint::endpoint()
    @        0x10c33878d ray::raylet::RayletConnection::RayletConnection()
    @        0x10c339e95 ray::raylet::RayletClient::RayletClient()
    @        0x10c29e0d7 ray::CoreWorker::CoreWorker()
    @        0x10c23c0d1 __pyx_tp_new_3ray_7_raylet_CoreWorker()
    @        0x10bbe9ae9 type_call
    @        0x10bbae72d _PyObject_FastCallKeywords
    @        0x10bc44977 call_function
    @        0x10bc3d5ef _PyEval_EvalFrameDefault
    @        0x10bc452db _PyEval_EvalCodeWithName
    @        0x10bbae88c _PyFunction_FastCallKeywords
    @        0x10bc4497e call_function
    @        0x10bc3d733 _PyEval_EvalFrameDefault
    @        0x10bc452db _PyEval_EvalCodeWithName
    @        0x10bbae88c _PyFunction_FastCallKeywords
    @        0x10bc4497e call_function
    @        0x10bc3d733 _PyEval_EvalFrameDefault
    @        0x10bc452db _PyEval_EvalCodeWithName
    @        0x10bc3ba15 PyEval_EvalCode
    @        0x10bc6a2ab run_mod
    @        0x10bc692c6 PyRun_FileExFlags
[1]    2316 abort      python raytest.py
```

System: MacOS 10.15.3, Python 3.7.0, ray: tested with both 0.8.2 and latest.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://ray.readthedocs.io/en/latest/installation.html).

**Comment by walterddr** [comment]
seems like this is a problem with the underlying boost requirement:
https://www.boost.org/doc/libs/1_58_0/libs/filesystem/doc/reference.html#long-path-warning

Not sure if there any quick fix. :-(

**Comment by DavidMChan** [comment]
From the duplicate referenced issue above:

Calling ray.init(temp_dir="string > 45 chars") causes a std::terminate due to the 107 character socket length limit in unix: http://man7.org/linux/man-pages/man7/unix.7.html

(see also: https://unix.stackexchange.com/q/367008)

**Comment by robertnishihara** [comment]
@mehrdadn If there isn't a good way to just make it work, then how about the following?

- In the C++ constructor, explicitly check the length and give a fatal error if it is too long.
- In Python and Java (in `ray.init()`), raise an exception if the name is too long?

**Comment by mehrdadn** [comment]
Yeah, something like that should work for now. The main complication here is—what length do we check it against? The viability of a `temp_dir` depends on the subsequent socket path names we want to use, which are different, and in multiple places in the code. We can glance through for the maximum length used right now, but it'll likely change over time. Getting it exactly right would need some refactoring, e.g. to let us aggregate all possible suffixes etc.

**Comment by walterddr** [comment]
+1 to @robertnishihara 's solution. not having ray crashing upon init (with explicit error message) is always a good experience
another is to shorten the generated path like `.../session_2020-...`, for example using epoch instead. but that reduces readability during debugging so I am actually not too fond about this.

**Comment by robertnishihara** [comment]
Error message fixed by https://github.com/ray-project/ray/pull/7951. It still fails, but now with a better Python exception.

**Comment by GoingMyWay** [comment]
The same issue has not been resolved yet.

**Comment by alexlenail** [comment]
@walterddr is there a way to configure ray to have shorter session names via ray.init() to avoid this error?

**Comment by DanielWicz** [comment]
Is there any plan to resolve it ?

**Comment by dburkhardt** [comment]
Yeah this is really annoying. Is there no underlying fix? It seems unhelpful not to be able to accept arbitrary paths for temporary directories.

**Comment by mehrdadn** [comment]
Hi folks, if you're expecting this to get fixed, I suggest opening a new issue.
(I'm not one of the maintainers right now—I just know closed issues on most projects are typically not seen as having work pending.)

**Comment by afogarty85** [comment]
While this is still an going issue, any comment on how I can regularly delete the contents of my _temp_dir? While training, checkpoints are being saved to the node (which has very limited disk space). I want to set _temp_dir to cloud storage but I am blocked by this issue.

**Comment by m-samland** [comment]
I am still encountering this problem.

**Comment by svduplessis** [comment]
Any feedback from the team at Ray? This is still an issue, especially on cluster systems where we are required to set a custom temporary directory. @robertnishihara @mehrdadn @walterddr
